proposals/0000-local-modules.rst

```diff
@@ -0,0 +1,1200 @@
+
+   ::
+
+     let
+       x = u
+       import module A
+     in
+     v
+
+   The environment is modified for ``u`` and ``v``.
+
+   Let ``E`` be the ambient QEnv. Let ``I`` be the interpretation
+   of ``impspec`` with respect to the environment ``strip(modids, E)``. (If ``impspec``
+   is missing, then let ``I = strip(modids, E)``.)
+   The declaration ``import module modids (impspec)`` modifies the ambient QEnv
+   to become ``E ∪ I``.
+
+   Note that the declaration form includes the word ``module`` to distinguish
+   it from a normal ``import`` which induces a dependency on another file. An
+   ``import module`` declaration cannot induce a dependency.
+
+   If the declaration adds no new bindings into the environment (because, for example,
+   the module name is not used to qualify any identifiers), it is a warning
+   ``-Wredundant-local-import``.
+
+   This declaration form is allowed only with ``-XLocalModules``.
+
+Optional Change Specifications
+------------------------------
+
+Each numbered item in this section can be considered separately.
+
+1. A local module declaration can omit the module name,
+   making an anonymous local module. The names exported
+   by an anonymous module are not added to the enclosing environment qualified,
+   as there is no name to qualify by. (More formally: merge the export
+   environment of the local module with the ambient environment.)
+
+   New BNF::
+
+     topdecl ::= ... | 'module' [ 'qualified' ] [ modid ] [ exports ] 'where' decls
+
+   It is an error to omit a module's name and
+   include the ``qualified`` keyword, unless there is also an empty export list.
+   (Anonymous qualified modules with an
+   empty export list but are still useful
+   as a way of declaring instances that use local definitions.)
+
+#. Introduce a new export item with the following BNF::
+
+     export ::= ... | 'module' modids '(' '..' ')'
+
+   This export item exports all qualified names in scope with the ``modids`` prefix,
+   stripped of that prefix. ::
+
+     interpretExpItem('module' modids '(' '..' ')', ambient_env)
+       = strip(modids, ambient_env)
+
+   This new form is really the non-qualified equivalent of the new export
+   item ``module qualified modids``. It needs the ``(..)`` syntax to differentiate
+   it from ``module modids``, which is designed to be backward-compatible.
+
+   There is no non-qualified equivalent of ``module qualified modids exports``, with
+   an explicit export list.
+   
+#. Every ``class``, ``data``, ``newtype``, ``data instance``, and ``newtype
+   instance`` declaration with an alphanumeric name implicitly creates a new local module. The name of
+   the local module matches the name of the declared type. All entities (e.g.,
+   method names, constructors, record selectors) brought into scope within the
+   declaration, including the type itself, are put into this local module.
+   
+   If the pseudo-keyword ``qualified`` appears directly after the keyword(s)
```

Comment:
Why “pseudo-keyword”? Does
```suggestion
   If the keyword ``qualified`` appears directly after the keyword(s)
```
change the meaning in some way?

Reply:
`qualified` is a pseudo-keyword in Haskell. That is, it has a special meaning in a particular context, but is not a `reservedid` of https://www.haskell.org/onlinereport/haskell2010/haskellch10.html#x17-17500010. I would interpret "keyword" here to suggest a change in this status.